demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb

```diff
@@ -0,0 +1,631 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 82,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "qa_chain = RetrievalQA.from_chain_type(\n",
+    "    llm=llm, \n",
+    "    retriever=retriever,     \n",
+    "    chain_type_kwargs={\n",
+    "        \"prompt\": PromptTemplate(\n",
+    "            template=template,\n",
+    "            input_variables=[\"context\", \"question\"],\n",
+    "        ),\n",
+    "    }\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we should have a working chain for QA. Let's test it out before wire it up with UI blocks."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "result = qa_chain({\"query\": \"Why choose Llama?\"})\n",
+    "print(result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "After confirming the validity, we can start building the UI. Before we define the gradio blocks, let's first define the callback streams that we will use later for streaming feature.  \n",
+    "This callback handler will put streaming LLM responses to a queue for gradio UI to render on the flight. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 84,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "job_done = object()\n",
+    "\n",
+    "class MyStream(StreamingStdOutCallbackHandler):\n",
+    "    def __init__(self, q) -> None:\n",
+    "        self.q = q\n",
+    "\n",
+    "    def on_llm_new_token(self, token: str, **kwargs: Any) -> None:\n",
+    "        self.q.put(token)\n",
+    "\n",
+    "    def on_llm_end(self, response: LLMResult, **kwargs: Any) -> None:\n",
+    "        self.q.put(job_done)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we can define the gradio UI blocks.  \n",
+    "Since we will need to define the UI and handlers in the same place, this will be a large chunk of code. We will add comments in the code for explanation."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 85,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import gradio as gr\n",
+    "\n",
+    "with gr.Blocks() as demo:\n",
+    "    #Configure UI layout\n",
+    "    chatbot = gr.Chatbot(height = 600)\n",
+    "    with gr.Row():\n",
+    "        with gr.Column(scale=1):\n",
+    "            with gr.Row():\n",
+    "                #model selection\n",
+    "                model_selector = gr.Dropdown(list(model_dict.keys()), value=\"7b-chat\", label=\"Model\", info=\"Select the model\", interactive = True, scale=1)\n",
+    "                max_new_tokens_selector = gr.Number(value=512, precision=0, label=\"Max new tokens\", info=\"Adjust max_new_tokens\",interactive = True, minimum=1, maximum=1024, scale=1)\n",
+    "            with gr.Row():\n",
+    "                #hyperparameter selection\n",
```

Comment:
hyperparameter selection -> temperature selection
as temperature is not really a hyperparameter.

Reply:
Hmm, I think it is. At least in our context or habit. [Reference1](https://medium.com/@dixnjakindah/top-p-temperature-and-other-parameters-1a53d2f8d7d7#:~:text=There%20are%20many%20other%20hyperparameters,time%2C%20accuracy%2C%20and%20generalizability.). [Reference2](https://deepchecks.com/glossary/llm-parameters/#:~:text=One%20intriguing%20parameter%20within%20LLMs,of%20straying%20from%20the%20context.)